x-pack/legacy/plugins/index_management/public/components/template_form/template_form.tsx

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import React, { Fragment, useState } from 'react';
+import { FormattedMessage } from '@kbn/i18n/react';
+import {
+  EuiButton,
+  EuiButtonEmpty,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiForm,
+  EuiSpacer,
+} from '@elastic/eui';
+import { Template } from '../../../common/types';
+import { TemplateSteps } from './template_steps';
+import { StepAliases, StepLogistics, StepMappings, StepSettings, StepReview } from './steps';
+import {
+  validateLogistics,
+  validateSettings,
+  validateMappings,
+  validateAliases,
+  TemplateValidation,
+} from '../../services/template_validation';
+import { StepProps } from './types';
+import { BASE_PATH } from '../../../common/constants';
+import { SectionError } from '..';
+
+interface Props {
+  onSave: (template: Template) => void;
+  clearSaveError: () => void;
+  isSaving: boolean;
+  saveError: any;
+  template: Template;
+  isEditing?: boolean;
+}
+
+const defaultValidation = {
+  isValid: true,
+  errors: {},
+};
+
+const stepComponentMap: { [key: number]: React.FunctionComponent<StepProps> } = {
+  1: StepLogistics,
+  2: StepSettings,
+  3: StepMappings,
+  4: StepAliases,
+  5: StepReview,
+};
+
+const stepValidationMap: { [key: number]: any } = {
+  1: validateLogistics,
+  2: validateSettings,
+  3: validateMappings,
+  4: validateAliases,
+};
+
+export const TemplateForm: React.FunctionComponent<Props> = ({
+  template: initialTemplate,
+  onSave,
+  isSaving,
+  saveError,
+  clearSaveError,
+  isEditing,
+}) => {
+  // hooks
+  const [currentStep, setCurrentStep] = useState<number>(1);
+  const [maxCompletedStep, setMaxCompletedStep] = useState<number>(0);
+  const [template, setTemplate] = useState<Template>(initialTemplate);
+  const [validation, setValidation] = useState<{ [key: number]: TemplateValidation }>({
+    1: defaultValidation,
+    2: defaultValidation,
+    3: defaultValidation,
+    4: defaultValidation,
+    5: defaultValidation,
+  });
+
+  const lastStep = Object.keys(stepComponentMap).length;
+
+  const CurrentStepComponent = stepComponentMap[currentStep];
+
+  const validateStep = stepValidationMap[currentStep];
+
+  const stepErrors = validation[currentStep].errors;
+  const isStepValid = validation[currentStep].isValid;
+
+  const validationErrors = Object.keys(stepErrors).reduce((acc: any, key: string) => {
+    return [...acc, ...stepErrors[key]];
+  }, []);
+
+  const updateTemplate = (updatedTemplate: Partial<Template>): void => {
+    const newTemplate = { ...template, ...updatedTemplate };
+
+    setTemplate(newTemplate);
+  };
+
+  const handleStepChange = (newCurrentStep: number, newMaxCompletedStep: number) => {
```

Comment:
I still think the logic could still be simplified. Aside, there is a small bug. If we go to step 2, insert an invalid JSON then click "back".  And then go the review step, the invalid JSON is sent over.

<img width="250" alt="Screen Shot 2019-08-22 at 09 06 50" src="https://user-images.githubusercontent.com/2854616/63497398-f172a500-c4c3-11e9-997e-ca58765aee11.png">

This simplified logic fixes this bug

```js
const updateCurrentStep = (nextStep: number) => {
  // All the steps needs validation, except for the fifth
  const shouldValidate = currentStep !== 5;

  if (shouldValidate) {
    const stepValidation = validateStep(template);

    const { isValid: isCurrentStepValid } = stepValidation;

    const newValidation = {
      ...validation,
      ...{
        [currentStep]: stepValidation,
      },
    };

    setValidation(newValidation);

    // If step is invalid do not let user to proceed
    if (!isCurrentStepValid) {
      return;
    }
  }

  setCurrentStep(nextStep);
  setMaxCompletedStep(prev => Math.max(nextStep - 1, prev));
  clearSaveError();
};

const onBack = () => {
  updateCurrentStep(currentStep - 1);
};

const onNext = () => {
  updateCurrentStep(currentStep + 1);
};
```

Reply:
good catch! the calculation for the max completed step differs for `onBack` vs `onNext`, so i tweaked this a little bit, but should be much cleaner now. thanks!

Reply:
Mmmm... why does it defer?

Reply:
Sorry, misspoke. The bug with the code above^ exists if you try to go **back > 1** step via the steps component. For example, if I'm on the `Aliases` step and click back to `Logistics`, this is what I see:

<img width="1299" alt="Screen Shot 2019-08-22 at 1 40 58 PM" src="https://user-images.githubusercontent.com/5226211/63537130-09a5ec80-c4e3-11e9-9a89-aceb61b20c65.png">

Reply:
Ah I see. Then we don't need this `maxCompletedStep` state. If we completely remove it, then the <TemplateSteps /> component would have this JSX

```js
const steps = [1, 2, 3, 4, 5].map(step => {
    return {
      title: stepNamesMap[step],
      isComplete: currentStep > step,
      isSelected: currentStep === step,
      onClick: () => updateCurrentStep(step, step - 1),
    };
  });
```